feat: Allow users to override the default deploy tool workspace directory

96malhar opened the pull request:
*Issue #, if available:*
DOTNET-6010

*Description of changes:*
The PR gives users the ability to set their custom deploy tool workspace directory. They can do so by setting a user scoped or machine scoped environment variable called `AWS_DOTNET_DEPLOYTOOL_WORKSPACE`. This variable must point to a directory that exists on the disk and does not contain any whitespace characters.

This PR also solves the issue of failed deployments when the user profile path contains a whitespace character - https://github.com/aws/aws-dotnet-deploy/issues/579
The underlying issue is with CDK and not with our deployment tool - https://github.com/aws/aws-cdk/issues/2532 and https://github.com/aws/jsii/issues/1296

I will create separate PR to update the documentation and the troubleshooting guide. I am unsure of how to cover this feature with an integration test as it would require `AWS_DOTNET_DEPLOYTOOL_WORKSPACE` to exist as a `User` or `Machine` scoped environment variable. But since our tests run in parallel, it will be caught up by other unintended integ tests too.

-----------

**Without the setting the `AWS_DOTNET_DEPLOYTOOL_WORKSPACE` environment variable**

![Capture](https://user-images.githubusercontent.com/36622308/173954888-dd263031-b294-423e-8f75-c78d8a1d5d38.PNG)
---------------

**The same deployment is successful after the `AWS_DOTNET_DEPLOYTOOL_WORKSPACE` environment variable**

![Capture](https://user-images.githubusercontent.com/36622308/173955117-c0500055-59ff-49b0-b491-0dacf0d856b8.PNG)
----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

normj commented on src/AWS.Deploy.Orchestration/Utilities/EnvironmentVariableManager.cs:
```diff
@@ -0,0 +1,36 @@
+
+    public class EnvironmentVariableManager : IEnvironmentVariableManager
+    {
+        public string? GetEnvironmentVariable(string variable, EnvironmentVariableTarget target)
```
I don't think we need to worry about `EnvironmentVariableTarget`. You can just call the `GetEnvironmentVariable` that doesn't take that variable which will work for both User and Machine. Otherwise when ever we call this we will have to call it for both user and machine.

96malhar replied:
Fixed

normj commented on src/AWS.Deploy.Orchestration/Utilities/Helpers.cs:
```diff
@@ -0,0 +1,90 @@
+        /// </summary>
+        public static string GetDeployToolWorkspaceDirectoryRoot(string userProfilePath, IDirectoryManager directoryManager, IEnvironmentVariableManager environmentVariableManager)
+        {
+            var overridenWorkspace = environmentVariableManager.GetEnvironmentVariable("AWS_DOTNET_DEPLOYTOOL_WORKSPACE", EnvironmentVariableTarget.User)
```
Use at least a local constant for `AWS_DOTNET_DEPLOYTOOL_WORKSPACE`.

96malhar replied:
Added a public constant.

normj commented on src/AWS.Deploy.Orchestration/Utilities/Helpers.cs:
```diff
@@ -0,0 +1,90 @@
+                return !defaultWorkSpace.Contains(" ")
+                    ? defaultWorkSpace
+                    : throw new InvalidDeployToolWorkspaceException(DeployToolErrorCode.InvalidDeployToolWorkspace, $"The USERPROFILE path ({userProfilePath}) contains a whitespace character and cannot be used as a workspace by the deployment tool. " +
+                    "Please refer to the troubleshooting guide for additional configuration required to handle this issue.");
```
Can we be more prescriptive here and tell them refer to the troubleshooting guide for setting the `AWS_DOTNET_DEPLOYTOOL_WORKSPACE` environment variable to use an alternative workspace error.

96malhar replied:
Done

normj commented on src/AWS.Deploy.Orchestration/Utilities/Helpers.cs:
```diff
@@ -0,0 +1,90 @@
+                    $"The AWS_DOTNET_DEPLOYTOOL_WORKSPACE environment variable ({overridenWorkspace}) contains a whitespace character and cannot be used as a workspace by the deployment tool.");
+            }
+
+            var tempDir = Path.Combine(overridenWorkspace, "temp");
```
Add comment about why we are setting the temp environment variable to work around CDK issue.

96malhar replied:
Done

normj reviewed (changes requested):
For testing since we have environment variables abstracted with `EnvironmentManager` I suggest adding a testing reset method on `DeployToolWorkspaceMetadata` that can inject a custom instance of `EnvironmentManager` with the workspace environment variable set.

96malhar commented:
> For testing since we have environment variables abstracted with `EnvironmentManager` I suggest adding a testing reset method on `DeployToolWorkspaceMetadata` that can inject a custom instance of `EnvironmentManager` with the workspace environment variable set.

Reworked the PR to use dependency injection and added an integ test.

normj reviewed (approved):
Just a minor comment about the doc update but otherwise I'm good with the change.
